File: .sling/libs/web5/eqela/jk-net/0.38.0/jk.web.json.js
let JkWebJsonJSONAPIClient = function() {
	this.credentialsEnabled = false;
	this.redirectionEnabled = false;
	this.redirectionLimit = 5;
	this.apiUrl = null;
	this.webClient = null;
	this.debugCallback = null;
};

JkWebJsonJSONAPIClient.prototype._t = { "JkWebJsonJSONAPIClient" : true };
JkWebJsonJSONAPIClient.prototype._tobj = JkWebJsonJSONAPIClient;

JkWebJsonJSONAPIClient.NEW = function() {
	var v = new JkWebJsonJSONAPIClient;
	return v.CTOR_JkWebJsonJSONAPIClient();
};

JkWebJsonJSONAPIClient.prototype.CTOR_JkWebJsonJSONAPIClient = function() {
	this.debugCallback = null;
	this.webClient = null;
	this.apiUrl = null;
	this.redirectionLimit = 5;
	this.redirectionEnabled = false;
	this.credentialsEnabled = false;
	return this;
};

JkWebJsonJSONAPIClient.prototype.getFullURL = function(api) {
	var url = this.apiUrl;
	if(JkLangString.isEmpty(url)) {
		url = "/";
	}
	if(url == "/") {
		if(JkLangString.startsWith(api, "/", 0)) {
			return api;
		}
		return (JkLangString.safeString(url)) + (JkLangString.safeString(api));
	}
	if(JkLangString.endsWith(url, "/")) {
		if(JkLangString.startsWith(api, "/", 0)) {
			return (JkLangString.safeString(url)) + (JkLangString.safeString((JkLangString.getEndOfString(api, 1))));
		}
		return (JkLangString.safeString(url)) + (JkLangString.safeString(api));
	}
	if(JkLangString.startsWith(api, "/", 0)) {
		return (JkLangString.safeString(url)) + (JkLangString.safeString(api));
	}
	return (JkLangString.safeString(url)) + ("/") + (JkLangString.safeString(api));
};

JkWebJsonJSONAPIClient.prototype.toUTF8Buffer = function(data) {
	if(!(data != null)) {
		return null;
	}
	return JkLangString.toUTF8Buffer((JkJsonJSONEncoder.encode(data, true, false)));
};

JkWebJsonJSONAPIClient.prototype.customizeRequestHeaders = function(headers) {
};

JkWebJsonJSONAPIClient.prototype.onStartSendRequest = function() {
};

JkWebJsonJSONAPIClient.prototype.onEndSendRequest = function() {
};

JkWebJsonJSONAPIClient.prototype.onDefaultErrorHandler = function(error) {
};

JkWebJsonJSONAPIClient.prototype.handleAsError = function(response, callback) {
	var error = this.toError(response);
	if(error != null) {
		this.onError1(error, callback);
		return true;
	}
	return false;
};

JkWebJsonJSONAPIClient.prototype.toError = function(response) {
	if(!(response != null)) {
		return JkLangError.forCode("noServerResponse", null);
	}
	if(response.getString("status", null) == "error") {
		var v = JkLangError.NEW();
		var code = response.getString("code", null);
		if(JkLangString.isEmpty(code)) {
			code = response.getString("error", null);
		}
		v.setCode(code);
		var detail = response.getString("detail", null);
		if(JkLangString.isEmpty(detail)) {
			detail = response.getString("message", null);
		}
		v.setDetail(detail);
		return v;
	}
	return null;
};

JkWebJsonJSONAPIClient.prototype.onError1 = function(error, callback) {
	if(!(callback != null)) {
		this.onDefaultErrorHandler(error);
		return;
	}
	callback(error);
};

JkWebJsonJSONAPIClient.prototype.onError2 = function(error) {
	this.onError1(error, null);
};

JkWebJsonJSONAPIClient.prototype.sendRequest = function(method, url, headers, data, callback, errorCallback) {
	if(!(callback != null)) {
		return;
	}
	var ll = callback;
	var ecb = errorCallback;
	this.doSendRequest(method, url, headers, data, (function(responseHeaders1, body1) {
		if(!(body1 != null && JkLangBuffer.getSize(body1) > 0)) {
			this.onError1((JkLangError.forCode("requestFailed", null)), ecb);
			return;
		}
		var responseString = JkLangString.forUTF8Buffer(body1);
		if(this.debugCallback != null) {
			this.debugCallback("Received response: `" + (JkLangString.safeString(responseString)) + ("'"));
		}
		var jsonResponseBody = (function(o) {
			if(JkLangDynamicMap.IS_INSTANCE && JkLangDynamicMap.IS_INSTANCE(o)) {
				return o;
			}
			return null;
		}.bind(this))((JkJsonJSONParser.parseString(responseString)));
		if(!(jsonResponseBody != null)) {
			this.onError1((JkLangError.forCode("invalidServerResponse", null)), ecb);
			return;
		}
		if(!this.handleAsError(jsonResponseBody, ecb)) {
			ll(jsonResponseBody);
		}
	}.bind(this)), ecb, 0);
};

JkWebJsonJSONAPIClient.prototype.doSendRequest = function(method, url, headers, data, callback, errorCallback, redirectCount) {
	if(!(callback != null)) {
		return;
	}
	if(!(redirectCount <= this.redirectionLimit)) {
		this.onError1((JkLangError.forCode("tooManyRedirections", null)), errorCallback);
		return;
	}
	var hrs = headers;
	if(!(hrs != null)) {
		hrs = JkLangKeyValueList.NEW();
		hrs.add("Content-Type", "application/json");
	}
	if(!(this.webClient != null)) {
		this.webClient = JkWebNativeWebClient.instance();
	}
	if(!(this.webClient != null)) {
		this.onError1((JkLangError.forCode("noWebClient", null)), errorCallback);
		return;
	}
	this.webClient.setCredentialsEnabled(this.credentialsEnabled);
	this.customizeRequestHeaders(hrs);
	var hrsf = hrs;
	var cb = callback;
	var ecb = errorCallback;
	var urlf = url;
	var dataf = data;
	var methodf = method;
	this.onStartSendRequest();
	if(this.debugCallback != null) {
		this.debugCallback("Sending request: `" + (JkLangString.safeString((JkLangString.forUTF8Buffer(data)))) + ("'"));
	}
	var rdc = redirectCount;
	this.webClient.query(methodf, urlf, hrsf, dataf, (function(statusCode1, responseHeaders1, body1) {
		this.onEndSendRequest();
		if(this.redirectionEnabled) {
			var code = JkLangString.toInteger(statusCode1);
			if(code >= 300 && code <= 399 && responseHeaders1 != null) {
				var location = null;
				var rhs = responseHeaders1.asVector();
				if(rhs != null) {
					var n2 = 0;
					var m = rhs.length;
					for(n2 = 0 ; n2 < m ; n2++) {
						var kvp = rhs[n2];
						if(kvp != null) {
							if(JkLangString.equalsIgnoreCase(kvp.key, "Location")) {
								location = kvp.value;
								break;
							}
						}
					}
				}
				if(JkLangString.isNotEmpty(location)) {
					if(!JkLangString.startsWith(location, "http", 0)) {
						location = this.getFullURL(location);
					}
					var o = JkUrlURL.forString(urlf, false);
					var n = JkUrlURL.forString(location, false);
					if(o != null && n != null && JkLangString.equalsIgnoreCase((o.getHost()), (n.getHost())) && !(o.getPort() == n.getPort())) {
						n.setPort((o.getPort()));
						location = n.toString();
					}
					this.doSendRequest(methodf, location, hrsf, dataf, cb, ecb, (rdc + 1));
					return;
				}
			}
		}
		if(!(body1 != null && JkLangBuffer.getSize(body1) > 0)) {
			this.onError1((JkLangError.forCode("failedToConnect", null)), ecb);
			return;
		}
		cb(responseHeaders1, body1);
	}.bind(this)));
};

JkWebJsonJSONAPIClient.prototype.doGetFile = function(url, callback, errorCallback) {
	this.doSendRequest("GET", (this.getFullURL(url)), null, null, callback, errorCallback, 0);
};

JkWebJsonJSONAPIClient.prototype.doGet = function(url, callback, errorCallback) {
	this.sendRequest("GET", (this.getFullURL(url)), null, null, callback, errorCallback);
};

JkWebJsonJSONAPIClient.prototype.doPost = function(url, data, callback, errorCallback) {
	this.sendRequest("POST", (this.getFullURL(url)), null, (this.toUTF8Buffer(data)), callback, errorCallback);
};

JkWebJsonJSONAPIClient.prototype.doPut = function(url, data, callback, errorCallback) {
	this.sendRequest("PUT", (this.getFullURL(url)), null, (this.toUTF8Buffer(data)), callback, errorCallback);
};

JkWebJsonJSONAPIClient.prototype.doDelete = function(url, callback, errorCallback) {
	this.sendRequest("DELETE", (this.getFullURL(url)), null, null, callback, errorCallback);
};

JkWebJsonJSONAPIClient.prototype.uploadFile = function(url, data, mimeType, callback, errorCallback) {
	var mt = mimeType;
	if(JkLangString.isEmpty(mt)) {
		mt = "application/octet-stream";
	}
	var hdrs = JkLangKeyValueList.NEW();
	hdrs.add("content-type", mt);
	this.sendRequest("POST", (this.getFullURL(url)), hdrs, data, callback, errorCallback);
};

JkWebJsonJSONAPIClient.prototype.getWithStatus = function(url, callback) {
	var cb = callback;
	var okcb = function(v1) {
		var data = v1.getDynamicMap("data");
		if(data == null) {
			data = JkLangDynamicMap.NEW();
		}
		cb(data, null);
	}.bind(this);
	var errcb = function(err1) {
		cb(null, err1);
	}.bind(this);
	this.sendRequest("GET", (this.getFullURL(url)), null, null, okcb, errcb);
};

JkWebJsonJSONAPIClient.prototype.postWithStatus = function(url, data1, headers, callback) {
	var cb = callback;
	var okcb = function(v1) {
		var data = v1.getDynamicMap("data");
		if(data == null) {
			data = JkLangDynamicMap.NEW();
		}
		cb(data, null);
	}.bind(this);
	var errcb = function(err1) {
		cb(null, err1);
	}.bind(this);
	this.sendRequest("POST", (this.getFullURL(url)), headers, (this.toUTF8Buffer(data1)), okcb, errcb);
};

JkWebJsonJSONAPIClient.prototype.getCredentialsEnabled = function() {
	return this.credentialsEnabled;
};

JkWebJsonJSONAPIClient.prototype.setCredentialsEnabled = function(v) {
	this.credentialsEnabled = v;
	return this;
};

JkWebJsonJSONAPIClient.prototype.getRedirectionEnabled = function() {
	return this.redirectionEnabled;
};

JkWebJsonJSONAPIClient.prototype.setRedirectionEnabled = function(v) {
	this.redirectionEnabled = v;
	return this;
};

JkWebJsonJSONAPIClient.prototype.getRedirectionLimit = function() {
	return this.redirectionLimit;
};

JkWebJsonJSONAPIClient.prototype.setRedirectionLimit = function(v) {
	this.redirectionLimit = v;
	return this;
};

JkWebJsonJSONAPIClient.prototype.getApiUrl = function() {
	return this.apiUrl;
};

JkWebJsonJSONAPIClient.prototype.setApiUrl = function(v) {
	this.apiUrl = v;
	return this;
};

JkWebJsonJSONAPIClient.prototype.getWebClient = function() {
	return this.webClient;
};

JkWebJsonJSONAPIClient.prototype.setWebClient = function(v) {
	this.webClient = v;
	return this;
};

JkWebJsonJSONAPIClient.prototype.getDebugCallback = function() {
	return this.debugCallback;
};

JkWebJsonJSONAPIClient.prototype.setDebugCallback = function(v) {
	this.debugCallback = v;
	return this;
};

JkWebJsonJSONAPIClient.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkWebJsonJSONAPIClient"] === true;
};

File: .sling/libs/sushi/eqela/jk-net/0.27.0/jk.web.json.lua
jk = jk or {}

jk.web = jk.web or {}

jk.web.json = jk.web.json or {}

jk.web.json.JSONAPIClient = {}
jk.web.json.JSONAPIClient.__index = jk.web.json.JSONAPIClient
_vm:set_metatable(jk.web.json.JSONAPIClient, {})

function jk.web.json.JSONAPIClient._create()
	local v = _vm:set_metatable({}, jk.web.json.JSONAPIClient)
	return v
end

function jk.web.json.JSONAPIClient:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.web.json.JSONAPIClient'
	self['_isType.jk.web.json.JSONAPIClient'] = true
	self.credentialsEnabled = false
	self.redirectionEnabled = false
	self.redirectionLimit = 5
	self.apiUrl = nil
	self.webClient = nil
	self.debugCallback = nil
end

function jk.web.json.JSONAPIClient:_construct0()
	jk.web.json.JSONAPIClient._init(self)
	return self
end

function jk.web.json.JSONAPIClient:getFullURL(api)
	local url = self.apiUrl
	if _g.jk.lang.String:isEmpty(url) then
		url = "/"
	end
	if url == "/" then
		if _g.jk.lang.String:startsWith(api, "/", 0) then
			do return api end
		end
		do return _g.jk.lang.String:safeString(url) .. _g.jk.lang.String:safeString(api) end
	end
	if _g.jk.lang.String:endsWith(url, "/") then
		if _g.jk.lang.String:startsWith(api, "/", 0) then
			do return _g.jk.lang.String:safeString(url) .. _g.jk.lang.String:safeString(_g.jk.lang.String:getEndOfString(api, 1)) end
		end
		do return _g.jk.lang.String:safeString(url) .. _g.jk.lang.String:safeString(api) end
	end
	if _g.jk.lang.String:startsWith(api, "/", 0) then
		do return _g.jk.lang.String:safeString(url) .. _g.jk.lang.String:safeString(api) end
	end
	do return _g.jk.lang.String:safeString(url) .. "/" .. _g.jk.lang.String:safeString(api) end
end

function jk.web.json.JSONAPIClient:toUTF8Buffer(data)
	if not (data ~= nil) then
		do return nil end
	end
	do return _g.jk.lang.String:toUTF8Buffer(_g.jk.json.JSONEncoder:encode(data, true, false)) end
end

function jk.web.json.JSONAPIClient:customizeRequestHeaders(headers)
end

function jk.web.json.JSONAPIClient:onStartSendRequest()
end

function jk.web.json.JSONAPIClient:onEndSendRequest()
end

function jk.web.json.JSONAPIClient:onDefaultErrorHandler(error)
end

function jk.web.json.JSONAPIClient:handleAsError(response, callback)
	local error = self:toError(response)
	if error ~= nil then
		do self:onError1(error, callback) end
		do return true end
	end
	do return false end
end

function jk.web.json.JSONAPIClient:toError(response)
	if not (response ~= nil) then
		do return _g.jk.lang.Error:forCode("noServerResponse", nil) end
	end
	if response:getString("status", nil) == "error" then
		local v = _g.jk.lang.Error._construct0(_g.jk.lang.Error._create())
		local code = response:getString("code", nil)
		if _g.jk.lang.String:isEmpty(code) then
			code = response:getString("error", nil)
		end
		do v:setCode(code) end
		do
			local detail = response:getString("detail", nil)
			if _g.jk.lang.String:isEmpty(detail) then
				detail = response:getString("message", nil)
			end
			do v:setDetail(detail) end
			do return v end
		end
	end
	do return nil end
end

function jk.web.json.JSONAPIClient:onError1(error, callback)
	if not (callback ~= nil) then
		do self:onDefaultErrorHandler(error) end
		do return end
	end
	do callback(error) end
end

function jk.web.json.JSONAPIClient:onError2(error)
	do self:onError1(error, nil) end
end

function jk.web.json.JSONAPIClient:sendRequest(method, url, headers, data, callback, errorCallback)
	if not (callback ~= nil) then
		do return end
	end
	do
		local ll = callback
		local ecb = errorCallback
		do self:doSendRequest(method, url, headers, data, function(responseHeaders, body)
			if not (body ~= nil and _g.jk.lang.Buffer:getSize(body) > 0) then
				do self:onError1(_g.jk.lang.Error:forCode("requestFailed", nil), ecb) end
				do return end
			end
			do
				local responseString = _g.jk.lang.String:forUTF8Buffer(body)
				if self.debugCallback ~= nil then
					do self.debugCallback("Received response: `" .. _g.jk.lang.String:safeString(responseString) .. "'") end
				end
				do
					local jsonResponseBody = _vm:to_table_with_key(_g.jk.json.JSONParser:parseString(responseString), '_isType.jk.lang.DynamicMap')
					if not (jsonResponseBody ~= nil) then
						do self:onError1(_g.jk.lang.Error:forCode("invalidServerResponse", nil), ecb) end
						do return end
					end
					if not self:handleAsError(jsonResponseBody, ecb) then
						do ll(jsonResponseBody) end
					end
				end
			end
		end, ecb, 0) end
	end
end

function jk.web.json.JSONAPIClient:doSendRequest(method, url, headers, data, callback, errorCallback, redirectCount)
	if not (callback ~= nil) then
		do return end
	end
	if not (redirectCount <= self.redirectionLimit) then
		do self:onError1(_g.jk.lang.Error:forCode("tooManyRedirections", nil), errorCallback) end
		do return end
	end
	do
		local hrs = headers
		if not (hrs ~= nil) then
			hrs = _g.jk.lang.KeyValueList._construct0(_g.jk.lang.KeyValueList._create())
			do hrs:add("Content-Type", "application/json") end
		end
		if not (self.webClient ~= nil) then
			self.webClient = _g.jk.web.NativeWebClient:instance()
		end
		if not (self.webClient ~= nil) then
			do self:onError1(_g.jk.lang.Error:forCode("noWebClient", nil), errorCallback) end
			do return end
		end
		do self.webClient:setCredentialsEnabled(self.credentialsEnabled) end
		do self:customizeRequestHeaders(hrs) end
		do
			local hrsf = hrs
			local cb = callback
			local ecb = errorCallback
			local urlf = url
			local dataf = data
			local methodf = method
			do self:onStartSendRequest() end
			if self.debugCallback ~= nil then
				do self.debugCallback("Sending request: `" .. _g.jk.lang.String:safeString(_g.jk.lang.String:forUTF8Buffer(data)) .. "'") end
			end
			do
				local rdc = redirectCount
				do self.webClient:query(methodf, urlf, hrsf, dataf, function(statusCode, responseHeaders, body)
					do self:onEndSendRequest() end
					if self.redirectionEnabled then
						local code = _g.jk.lang.String:toInteger(statusCode)
						if code >= 300 and code <= 399 and responseHeaders ~= nil then
							local location = nil
							local rhs = responseHeaders:asVector()
							if rhs ~= nil then
								local n2 = 0
								local m = _g.jk.lang.Vector:getSize(rhs)
								do
									n2 = 0
									while n2 < m do
										local kvp = rhs[n2 + 1]
										if kvp ~= nil then
											if _g.jk.lang.String:equalsIgnoreCase(kvp.key, "Location") then
												location = kvp.value
												do break end
											end
										end
										do n2 = n2 + 1 end
									end
								end
							end
							if _g.jk.lang.String:isNotEmpty(location) then
								if not _g.jk.lang.String:startsWith(location, "http", 0) then
									location = self:getFullURL(location)
								end
								do
									local o = _g.jk.url.URL:forString(urlf, false)
									local n = _g.jk.url.URL:forString(location, false)
									if o ~= nil and n ~= nil and _g.jk.lang.String:equalsIgnoreCase(o:getHost(), n:getHost()) and not (o:getPort() == n:getPort()) then
										do n:setPort(o:getPort()) end
										location = n:toString()
									end
									do self:doSendRequest(methodf, location, hrsf, dataf, cb, ecb, rdc + 1) end
									do return end
								end
							end
						end
					end
					if not (body ~= nil and _g.jk.lang.Buffer:getSize(body) > 0) then
						do self:onError1(_g.jk.lang.Error:forCode("failedToConnect", nil), ecb) end
						do return end
					end
					do cb(responseHeaders, body) end
				end) end
			end
		end
	end
end

function jk.web.json.JSONAPIClient:doGetFile(url, callback, errorCallback)
	do self:doSendRequest("GET", self:getFullURL(url), nil, nil, callback, errorCallback, 0) end
end

function jk.web.json.JSONAPIClient:doGet(url, callback, errorCallback)
	do self:sendRequest("GET", self:getFullURL(url), nil, nil, callback, errorCallback) end
end

function jk.web.json.JSONAPIClient:doPost(url, data, callback, errorCallback)
	do self:sendRequest("POST", self:getFullURL(url), nil, self:toUTF8Buffer(data), callback, errorCallback) end
end

function jk.web.json.JSONAPIClient:doPut(url, data, callback, errorCallback)
	do self:sendRequest("PUT", self:getFullURL(url), nil, self:toUTF8Buffer(data), callback, errorCallback) end
end

function jk.web.json.JSONAPIClient:doDelete(url, callback, errorCallback)
	do self:sendRequest("DELETE", self:getFullURL(url), nil, nil, callback, errorCallback) end
end

function jk.web.json.JSONAPIClient:uploadFile(url, data, mimeType, callback, errorCallback)
	local mt = mimeType
	if _g.jk.lang.String:isEmpty(mt) then
		mt = "application/octet-stream"
	end
	do
		local hdrs = _g.jk.lang.KeyValueList._construct0(_g.jk.lang.KeyValueList._create())
		do hdrs:add("content-type", mt) end
		do self:sendRequest("POST", self:getFullURL(url), hdrs, data, callback, errorCallback) end
	end
end

function jk.web.json.JSONAPIClient:getWithStatus(url, callback)
	local cb = callback
	local okcb = function(v)
		local data = v:getDynamicMap("data")
		if data == nil then
			data = _g.jk.lang.DynamicMap._construct0(_g.jk.lang.DynamicMap._create())
		end
		do cb(data, nil) end
	end
	local errcb = function(err)
		do cb(nil, err) end
	end
	do self:sendRequest("GET", self:getFullURL(url), nil, nil, okcb, errcb) end
end

function jk.web.json.JSONAPIClient:postWithStatus(url, data, headers, callback)
	local cb = callback
	local okcb = function(v)
		local data = v:getDynamicMap("data")
		if data == nil then
			data = _g.jk.lang.DynamicMap._construct0(_g.jk.lang.DynamicMap._create())
		end
		do cb(data, nil) end
	end
	local errcb = function(err)
		do cb(nil, err) end
	end
	do self:sendRequest("POST", self:getFullURL(url), headers, self:toUTF8Buffer(data), okcb, errcb) end
end

function jk.web.json.JSONAPIClient:getCredentialsEnabled()
	do return self.credentialsEnabled end
end

function jk.web.json.JSONAPIClient:setCredentialsEnabled(v)
	self.credentialsEnabled = v
	do return self end
end

function jk.web.json.JSONAPIClient:getRedirectionEnabled()
	do return self.redirectionEnabled end
end

function jk.web.json.JSONAPIClient:setRedirectionEnabled(v)
	self.redirectionEnabled = v
	do return self end
end

function jk.web.json.JSONAPIClient:getRedirectionLimit()
	do return self.redirectionLimit end
end

function jk.web.json.JSONAPIClient:setRedirectionLimit(v)
	self.redirectionLimit = v
	do return self end
end

function jk.web.json.JSONAPIClient:getApiUrl()
	do return self.apiUrl end
end

function jk.web.json.JSONAPIClient:setApiUrl(v)
	self.apiUrl = v
	do return self end
end

function jk.web.json.JSONAPIClient:getWebClient()
	do return self.webClient end
end

function jk.web.json.JSONAPIClient:setWebClient(v)
	self.webClient = v
	do return self end
end

function jk.web.json.JSONAPIClient:getDebugCallback()
	do return self.debugCallback end
end

function jk.web.json.JSONAPIClient:setDebugCallback(v)
	self.debugCallback = v
	do return self end
end
